crates/mysten-network/src/callback/service.rs

```diff
@@ -0,0 +1,77 @@
+use super::{CallbackLayer, MakeCallbackHandler, ResponseFuture};
+use http::{Request, Response};
+use std::task::{Context, Poll};
+use tower::Service;
+
+/// Middleware that adds callbacks to a [`Service`].
+///
+/// See the [module docs](crate::callback) for an example.
+///
+/// [`Service`]: tower::Service
+#[derive(Debug, Clone, Copy)]
+pub struct Callback<S, M> {
+    pub(crate) inner: S,
+    pub(crate) make_callback_handler: M,
+}
+
+impl<S, M> Callback<S, M> {
+    /// Create a new [`Callback`].
+    pub fn new(inner: S, make_callback_handler: M) -> Self {
+        Self {
+            inner,
+            make_callback_handler,
+        }
+    }
+
+    /// Returns a new [`Layer`] that wraps services with a [`CallbackLayer`] middleware.
+    ///
+    /// [`Layer`]: tower::layer::Layer
+    pub fn layer(make_handler: M) -> CallbackLayer<M>
+    where
+        M: MakeCallbackHandler,
+    {
+        CallbackLayer::new(make_handler)
+    }
+}
+
+impl<S, M> Callback<S, M> {
```

Comment:
What is the advantage of having two impl for the same type?

Reply:
A Layer type and Service type are two different things. You can look at the tower ecosystem for more info on this.

Reply:
I am likely misunderstanding things, but line 17 and line 37 look the same to me? And the functions following the impl block on line 37 do not seem to use the M type. How should I interpret these two impl blocks?

Reply:
Ah I see what you mean. No there's no reason

Reply:
Fixed in the latest version